plugins/modules/nd_setup.py

```diff
@@ -442,6 +447,14 @@ def main():
     if state == "query":
         nd.existing = nd.request("/clusterstatus/install", method="GET")
     else:
+        nd_version = nd.query_obj("/version.json")
+        nd_version = ".".join(str(nd_version[key]) for key in ["major", "minor", "maintenance"])
+        if nd_version >= "3.0.1":
+            if not all((app_network, service_network)) and not all((app_network_ipv6, service_network_ipv6)):
+                nd.fail_json(msg="Application and service network addresses, IPv4 or IPv6, are required during ND setup.")
+        else:
+            if not all((app_network, service_network)):
+                nd.fail_json(msg="Application and service network IPv4 addresses are required during ND setup.")
```

Comment:
Shall we move check_network_requirements() outside the payload initialization for consistency? We usually validate attributes before setting the payload? The older payload will remain as is which is more readable?

```
for node in nodes:
def check_network_requirements(node, nd_version):
        network = node.get('management_network', {})
        if nd_version >= "3.0.1":
            if not all(network.get(ip) for ip in ["ipv6Subnet", "gatewayv6"]) and not all(network.get(ip) for ip in ["ipSubnet", "gateway"]):
                nd.fail_json(msg="Missing IPv4 subnet/gateway or IPv6 subnet/gateway in node {0} configuration.".format(type))
        else:
            if not all(network.get(ip) for ip in ["ipSubnet", "gateway"]):
                nd.fail_json(msg="Missing IPv4 subnet/gateway in node {0} configuration.".format(type))
```

Reply:
I moved it back before payload initialization as you suggested.